### PR TITLE
Founounou Homework 7

### PR DESCRIPTION
**Pull request by YassineFnn**
The second task is in monads/lambda.txt

### CODE REVIEW

File: monads/test/Spec.hs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+genInt = Gen.int (Range.constant 0 100)
+
+-- Generate expressions using only predefined variables from the initial context
+genExpr :: Int -> [String] -> Gen (Expr String)
+genExpr 0 boundVars =
+  if null boundVars
+    then Num <$> genInt
+    else Gen.choice [Num <$> genInt, Expr.Var <$> Gen.element boundVars]
+
+genExpr n boundVars = do
```
Comment (kajigor) [inline monads/test/Spec.hs]: It's better to use `Gen.recursive`: https://hackage.haskell.org/package/hedgehog-1.5/docs/Hedgehog-Gen.html#v:recursive

File: monads/test/Spec.hs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  case execProgram compiled initialState of
+    Left (VarUndefined variable) -> assert (variable == var_name)
+    _ -> fail "Expected VarUndefined error"
+
+prop_compile_stackUnderflow :: Property
+prop_compile_stackUnderflow = property $ do
+  let prog = [Add]
+  case execProgram prog initialState of
+    Left (StackUnderflow _) -> success
+    _ -> fail "Expected StackUnderflow error"
```
Comment (kajigor) [inline monads/test/Spec.hs]: This is just a unit test.

File: monads/test/Spec.hs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+prop_compile_stackNotExhausted :: Property
+prop_compile_stackNotExhausted = property $ do
+  int1 <- forAll genInt
+  int2 <- forAll genInt
+  let prog = [PushNum int1, PushNum int2]
+  case execProgram prog initialState of
+    Left (StackNotExhausted _) -> success
+    Right _ -> fail "Expected StackNotExhausted error due to extra items on the stack"
+    Left err -> fail $ "Unexpected error: " ++ show err
```
Comment (kajigor) [inline monads/test/Spec.hs]: This is also more of a unit test.

File: monads/test/Spec.hs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+prop_compile_varUndefined :: Property
+prop_compile_varUndefined = property $ do
+  var_name <- forAll $ Gen.string (Range.singleton 1) Gen.lower
+  let expr :: Expr String
+      expr = Expr.Var var_name
+  let compiled = compile expr
+  case execProgram compiled initialState of
+    Left (VarUndefined variable) -> assert (variable == var_name)
+    _ -> fail "Expected VarUndefined error"
```
Comment (kajigor) [inline monads/test/Spec.hs]: I would be more meaningful to add an undefined variable somewhere inside a random expression.

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
2 ok

File: monads/test/Spec.hs (hunk 1 above)
Comment (YassineFnn) [reply monads/test/Spec.hs]: Fixed.

File: monads/test/Spec.hs (hunk 2 above)
Comment (YassineFnn) [reply monads/test/Spec.hs]: Fixed

File: monads/test/Spec.hs (hunk 3 above)
Comment (YassineFnn) [reply monads/test/Spec.hs]: Fixed

File: monads/test/Spec.hs (hunk 4 above)
Comment (YassineFnn) [reply monads/test/Spec.hs]: Fixed

**Comment by kajigor** [comment]
ok

**Pull request closed without merging**